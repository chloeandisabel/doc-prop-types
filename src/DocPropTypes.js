const docProp = id => {
  const docProp = () => {
    return { id };
  };
  docProp.isRequired = () => {
    return { id: `${id}_IS_REQUIRED` };
  };
  return docProp;
};

const docPropWithArgs = id => {
  const docProp = args => {
    const reallyDocProp = () => {
      return { id, args };
    };
    reallyDocProp.isRequired = () => {
      return {
        id: `${id}_IS_REQUIRED`,
        args,
      };
    };
    return reallyDocProp;
  };

  return docProp;
};

export default {
  string: docProp("STRING"),
  number: docProp("NUMBER"),
  bool: docProp("BOOL"),
  symbol: docProp("SYMBOL"),
  func: docProp("FUNC"),
  object: docProp("OBJECT"),
  node: docProp("NODE"),
  element: docProp("ELEMENT"),
  array: docProp("ARRAY"),
  any: docProp("ANY"),
  shape: docPropWithArgs("SHAPE_OF"),
  oneOf: docPropWithArgs("ONE_OF"),
  oneOfType: docPropWithArgs("ONE_OF_TYPE"),
  arrayOf: docPropWithArgs("ARRAY_OF"),
  objectOf: docPropWithArgs("OBJECT_OF"),
  instanceOf: docPropWithArgs("INSTANCE_OF"),
  custom: docPropWithArgs("CUSTOM"),
};
