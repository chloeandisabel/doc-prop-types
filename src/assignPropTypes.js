import PropTypes from "prop-types";
import PropInfo from "./PropInfo";

export const optionalSimpleType = propTypeName => ({
  propInfo: PropInfo[propTypeName](),
  propType: PropTypes[propTypeName],
});

export const requiredSimpleType = propTypeName => ({
  propInfo: PropInfo[propTypeName].isRequired(),
  propType: PropTypes[propTypeName].isRequired,
});

export const optionalComplexType = (descriptor, primitive) => {
  const { propInfo, propType } = primitive(descriptor);

  return {
    propInfo: propInfo(),
    propType,
  };
};

export const requiredComplexType = (descriptor, primitive) => {
  const { propInfo, propType } = primitive(descriptor);

  return {
    propInfo: propInfo.isRequired(),
    propType: propType.isRequired,
  };
};

export const shapePrimitive = descriptor => ({
  propInfo: PropInfo.shape(
    Object.keys(descriptor.args).reduce((shape, key) => {
      shape[key] = getType(descriptor.args[key]()).propInfo;
      return shape;
    }, {}),
  ),
  propType: PropTypes.shape(
    Object.keys(descriptor.args).reduce((args, key) => {
      args[key] = getType(descriptor.args[key]()).propType;
      return args;
    }, {}),
  ),
});

export const oneOfPrimitive = descriptor => ({
  propInfo: PropInfo.oneOf(descriptor.args),
  propType: PropTypes.oneOf(descriptor.args),
});

export const oneOfTypePrimitive = descriptor => ({
  propInfo: PropInfo.oneOfType(
    descriptor.args.map(docPropType => {
      return getType(docPropType()).propInfo;
    }),
  ),
  propType: PropTypes.oneOfType(
    descriptor.args.map(docPropType => {
      return getType(docPropType()).propType;
    }),
  ),
});

export const arrayOfPrimitive = descriptor => ({
  propInfo: PropInfo.arrayOf(getType(descriptor.args()).propInfo),
  propType: PropTypes.arrayOf(getType(descriptor.args()).propType),
});

export const objectOfPrimitive = descriptor => ({
  propInfo: PropInfo.objectOf(getType(descriptor.args()).propInfo),
  propType: PropTypes.objectOf(getType(descriptor.args()).propType),
});

export const instanceOfPrimitive = docPropType => ({
  propInfo: PropInfo.instanceOf(docPropType.args),
  propType: PropTypes.instanceOf(docPropType.args),
});

export const customPrimitive = descriptor => ({
  propInfo: PropInfo.custom(descriptor.args),
  propType: descriptor.args,
});

export const getType = descriptor => {
  switch (descriptor.id) {
    case "STRING":
      return optionalSimpleType("string");

    case "NUMBER":
      return optionalSimpleType("number");

    case "BOOL":
      return optionalSimpleType("bool");

    case "SYMBOL":
      return optionalSimpleType("symbol");

    case "FUNC":
      return optionalSimpleType("func");

    case "OBJECT":
      return optionalSimpleType("object");

    case "NODE":
      return optionalSimpleType("node");

    case "ELEMENT":
      return optionalSimpleType("element");

    case "ARRAY":
      return optionalSimpleType("array");

    case "ANY":
      return optionalSimpleType("any");

    case "STRING_IS_REQUIRED":
      return requiredSimpleType("string");

    case "NUMBER_IS_REQUIRED":
      return requiredSimpleType("number");

    case "BOOL_IS_REQUIRED":
      return requiredSimpleType("bool");

    case "SYMBOL_IS_REQUIRED":
      return requiredSimpleType("symbol");

    case "FUNC_IS_REQUIRED":
      return requiredSimpleType("func");

    case "OBJECT_IS_REQUIRED":
      return requiredSimpleType("object");

    case "NODE_IS_REQUIRED":
      return requiredSimpleType("node");

    case "ELEMENT_IS_REQUIRED":
      return requiredSimpleType("element");

    case "ANY_IS_REQUIRED":
      return requiredSimpleType("any");

    case "SHAPE_OF":
      return optionalComplexType(descriptor, shapePrimitive);

    case "SHAPE_OF_IS_REQUIRED":
      return requiredComplexType(descriptor, shapePrimitive);

    case "ONE_OF":
      return optionalComplexType(descriptor, oneOfPrimitive);

    case "ONE_OF_IS_REQUIRED":
      return requiredComplexType(descriptor, oneOfPrimitive);

    case "ONE_OF_TYPE":
      return optionalComplexType(descriptor, oneOfTypePrimitive);

    case "ONE_OF_TYPE_IS_REQUIRED":
      return requiredComplexType(descriptor, oneOfTypePrimitive);

    case "ARRAY_OF":
      return optionalComplexType(descriptor, arrayOfPrimitive);

    case "ARRAY_OF_IS_REQUIRED":
      return requiredComplexType(descriptor, arrayOfPrimitive);

    case "OBJECT_OF":
      return optionalComplexType(descriptor, objectOfPrimitive);

    case "OBJECT_OF_IS_REQUIRED":
      return requiredComplexType(descriptor, objectOfPrimitive);

    case "INSTANCE_OF":
      return optionalComplexType(descriptor, instanceOfPrimitive);

    case "INSTANCE_OF_IS_REQUIRED":
      return requiredComplexType(descriptor, instanceOfPrimitive);

    case "CUSTOM":
      return optionalComplexType(descriptor, customPrimitive);

    case "CUSTOM_IS_REQUIRED":
      return requiredComplexType(descriptor, customPrimitive);

    default:
      /* eslint-disable no-console */
      console.error("An unexpected value was passed to assignPropTypes:");
      console.error(descriptor);
      console.error(
        "Make sure that youâ€™re using DocPropTypes, not React.PropTypes!",
      );
      /* eslint-enable no-console*/
      return {
        docPropType: null,
        propType: null,
      };
  }
};

export default function attachPropInfo(component, docPropTypes) {
  const propsWithInfo = Object.keys(docPropTypes).reduce(
    (propsWithInfo, key) => ({
      ...propsWithInfo,
      [key]: getType(docPropTypes[key]()),
    }),
    {},
  );

  component.propInfo = Object.keys(propsWithInfo).reduce(
    (propInfo, key) => ({
      ...propInfo,
      [key]: propsWithInfo[key].propInfo,
    }),
    {},
  );

  component.propTypes = Object.keys(propsWithInfo).reduce(
    (propTypes, key) => ({
      ...propTypes,
      [key]: propsWithInfo[key].propType,
    }),
    {},
  );

  component.docPropTypes = docPropTypes;

  return component;
}
