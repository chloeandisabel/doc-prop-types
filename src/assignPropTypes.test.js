import React from "react";
import assignPropTypes, {
  optionalSimpleType,
  requiredSimpleType,
  optionalComplexType,
  requiredComplexType,
  shapePrimitive,
  oneOfPrimitive,
  oneOfTypePrimitive,
  arrayOfPrimitive,
  objectOfPrimitive,
  instanceOfPrimitive,
  customPrimitive,
} from "./assignPropTypes";
import DocPropTypes from "./DocPropTypes";

describe("optionalSimpleType", () => {
  it("returns an object with propInfo and propType properties", () => {
    expect(optionalSimpleType("string")).toMatchSnapshot();
  });
});

describe("requiredSimpleType", () => {
  it("returns an object with required propInfo and propType properties", () => {
    expect(requiredSimpleType("number")).toMatchSnapshot();
  });
});

describe("optionalComplexType", () => {
  it("returns a complex prop with required set to false", () => {
    const descriptor = DocPropTypes.shape({
      name: DocPropTypes.string,
      age: DocPropTypes.number,
      pet: DocPropTypes.shape({
        isWellBehaved: DocPropTypes.bool,
      }),
    })();

    const actual = optionalComplexType(descriptor, shapePrimitive);

    expect(actual).toMatchSnapshot();
  });
});

describe("requiredComplexType", () => {
  it("returns a complex prop with required set to true", () => {
    const descriptor = DocPropTypes.shape({
      name: DocPropTypes.string,
      age: DocPropTypes.number,
      pet: DocPropTypes.shape({
        isWellBehaved: DocPropTypes.bool,
      }),
    })();

    const actual = requiredComplexType(descriptor, shapePrimitive);

    expect(actual).toMatchSnapshot();
  });
});

describe("shapePrimitive", () => {
  it("returns an object with info describing an object", () => {
    const descriptor = DocPropTypes.shape({
      name: DocPropTypes.string,
      age: DocPropTypes.number,
      pet: DocPropTypes.shape({
        isWellBehaved: DocPropTypes.bool,
      }),
    })();

    const actual = shapePrimitive(descriptor);

    expect(actual.propInfo()).toMatchSnapshot();
  });
});

describe("oneOfPrimitive", () => {
  it("returns an object with info describing an enum", () => {
    const descriptor = DocPropTypes.oneOf(["One", "Two", "Three"])();

    const actual = oneOfPrimitive(descriptor);

    expect(actual.propInfo()).toMatchSnapshot();
  });
});

describe("oneOfTypePrimitive", () => {
  it("returns an object with info describing a union type", () => {
    const descriptor = DocPropTypes.oneOfType([
      DocPropTypes.string,
      DocPropTypes.number.isRequired,
      DocPropTypes.bool,
    ])();

    const actual = oneOfTypePrimitive(descriptor);

    expect(actual.propInfo()).toMatchSnapshot();
  });
});

describe("arrayOfPrimitive", () => {
  it("returns an object with info describing an array with members of one type", () => {
    const descriptor = DocPropTypes.arrayOf(DocPropTypes.number)();
    const actual = arrayOfPrimitive(descriptor);

    expect(actual.propInfo()).toMatchSnapshot();
  });
});

describe("objectOfPrimitive", () => {
  it("returns an object with info describing an object with values of a given type", () => {
    const descriptor = DocPropTypes.objectOf(DocPropTypes.bool)();
    const actual = objectOfPrimitive(descriptor);

    expect(actual.propInfo()).toMatchSnapshot();
  });
});

describe("instanceOfPrimitive", () => {
  it("returns an object with info describing an instance of a class", () => {
    const descriptor = DocPropTypes.instanceOf(Date)();
    const actual = instanceOfPrimitive(descriptor);

    expect(actual.propInfo()).toMatchSnapshot();
  });
});

describe("customPrimitive", () => {
  it("returns an object with info describing a custom proptype", () => {
    const descriptor = DocPropTypes.custom(() => {})();
    const actual = customPrimitive(descriptor);

    expect(actual.propInfo()).toMatchSnapshot();
  });
});

describe("assignPropTypes", () => {
  it("assigns docPropTypes, propTypes, and propInfo to a component", () => {
    const component = () => React.createElement("div", [], []);

    const proppedComponent = assignPropTypes(component, {
      name: DocPropTypes.string.isRequired,
      age: DocPropTypes.number.isRequired,
      pet: DocPropTypes.shape({
        name: DocPropTypes.string.isRequired,
        isGoodBoy: DocPropTypes.bool.isRequired,
      }),
    });

    expect(proppedComponent.docPropTypes).toMatchSnapshot();
    expect(proppedComponent.propTypes).toMatchSnapshot();
    expect(proppedComponent.propInfo).toMatchSnapshot();
  });
});
