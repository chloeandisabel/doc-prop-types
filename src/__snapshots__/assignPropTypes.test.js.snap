// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arrayOfPrimitive returns an object with info describing an array with members of one type 1`] = `
Object {
  "isRequired": false,
  "memberType": Object {
    "isRequired": false,
    "type": "number",
  },
  "type": "arrayOf",
}
`;

exports[`assignPropTypes assigns docPropTypes, propTypes, and propInfo to a component 1`] = `
Object {
  "age": [Function],
  "name": [Function],
  "pet": [Function],
}
`;

exports[`assignPropTypes assigns docPropTypes, propTypes, and propInfo to a component 2`] = `
Object {
  "age": [Function],
  "name": [Function],
  "pet": [Function],
}
`;

exports[`assignPropTypes assigns docPropTypes, propTypes, and propInfo to a component 3`] = `
Object {
  "age": Object {
    "isRequired": true,
    "type": "number",
  },
  "name": Object {
    "isRequired": true,
    "type": "string",
  },
  "pet": Object {
    "isRequired": false,
    "shape": Object {
      "isGoodBoy": Object {
        "isRequired": true,
        "type": "bool",
      },
      "name": Object {
        "isRequired": true,
        "type": "string",
      },
    },
    "type": "shape",
  },
}
`;

exports[`customPrimitive returns an object with info describing a custom proptype 1`] = `
Object {
  "customPropType": [Function],
  "isRequired": false,
  "type": "custom",
}
`;

exports[`instanceOfPrimitive returns an object with info describing an instance of a class 1`] = `
Object {
  "instanceClass": [Function],
  "isRequired": false,
  "type": "instanceOf",
}
`;

exports[`objectOfPrimitive returns an object with info describing an object with values of a given type 1`] = `
Object {
  "isRequired": false,
  "type": "objectOf",
  "valueType": Object {
    "isRequired": false,
    "type": "bool",
  },
}
`;

exports[`oneOfPrimitive returns an object with info describing an enum 1`] = `
Object {
  "enums": Array [
    "One",
    "Two",
    "Three",
  ],
  "isRequired": false,
  "type": "oneOf",
}
`;

exports[`oneOfTypePrimitive returns an object with info describing a union type 1`] = `
Object {
  "isRequired": false,
  "type": "oneOfType",
  "types": Array [
    Object {
      "isRequired": false,
      "type": "string",
    },
    Object {
      "isRequired": true,
      "type": "number",
    },
    Object {
      "isRequired": false,
      "type": "bool",
    },
  ],
}
`;

exports[`optionalComplexType returns a complex prop with required set to false 1`] = `
Object {
  "propInfo": Object {
    "isRequired": false,
    "shape": Object {
      "age": Object {
        "isRequired": false,
        "type": "number",
      },
      "name": Object {
        "isRequired": false,
        "type": "string",
      },
      "pet": Object {
        "isRequired": false,
        "shape": Object {
          "isWellBehaved": Object {
            "isRequired": false,
            "type": "bool",
          },
        },
        "type": "shape",
      },
    },
    "type": "shape",
  },
  "propType": [Function],
}
`;

exports[`optionalSimpleType returns an object with propInfo and propType properties 1`] = `
Object {
  "propInfo": Object {
    "isRequired": false,
    "type": "string",
  },
  "propType": [Function],
}
`;

exports[`requiredComplexType returns a complex prop with required set to true 1`] = `
Object {
  "propInfo": Object {
    "isRequired": true,
    "shape": Object {
      "age": Object {
        "isRequired": false,
        "type": "number",
      },
      "name": Object {
        "isRequired": false,
        "type": "string",
      },
      "pet": Object {
        "isRequired": false,
        "shape": Object {
          "isWellBehaved": Object {
            "isRequired": false,
            "type": "bool",
          },
        },
        "type": "shape",
      },
    },
    "type": "shape",
  },
  "propType": [Function],
}
`;

exports[`requiredSimpleType returns an object with required propInfo and propType properties 1`] = `
Object {
  "propInfo": Object {
    "isRequired": true,
    "type": "number",
  },
  "propType": [Function],
}
`;

exports[`shapePrimitive returns an object with info describing an object 1`] = `
Object {
  "isRequired": false,
  "shape": Object {
    "age": Object {
      "isRequired": false,
      "type": "number",
    },
    "name": Object {
      "isRequired": false,
      "type": "string",
    },
    "pet": Object {
      "isRequired": false,
      "shape": Object {
        "isWellBehaved": Object {
          "isRequired": false,
          "type": "bool",
        },
      },
      "type": "shape",
    },
  },
  "type": "shape",
}
`;
